picoquic/tls_api.c

```diff
@@ -429,6 +429,77 @@ uint64_t picoquic_get_simulated_time_cb(ptls_get_time_t* self)
     return ((**pp_simulated_time) / 1000);
 }
 
+typedef struct {
+    ptls_verify_certificate_t cb;
+    picoquic_quic_t *quic;
+} picoquic_verify_certificate_t;
+
+typedef struct {
+    /* The pointer to the overlying `verify_ctx` */
+    void *verify_ctx;
+    int (*verify_sign)(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign);
+} picoquic_verify_ctx_t;
+
+static int verify_sign_callback(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign)
+{
+    picoquic_verify_ctx_t* ctx = (picoquic_verify_ctx_t*)verify_ctx;
+    int ret = 0;
+
+    ret = ctx->verify_sign(ctx->verify_ctx, data, sign);
+
+    free(ctx);
+
+    return ret;
+}
+
+#define container_of(ptr, type, member) ((type *)((char *)(ptr) - offsetof(type, member)))
+
+static int verify_certificate_callback(ptls_verify_certificate_t* _self, void* ctx, ptls_t *tls,
+                                       int (**verify_sign)(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign),
+                                       void **verify_data,
+                                       ptls_iovec_t *certs,
+                                       size_t num_certs)
+{
+    picoquic_verify_certificate_t *self = container_of(_self, picoquic_verify_certificate_t, cb);
+    picoquic_cnx_t* cnx = (picoquic_cnx_t*)ctx;
+    int ret = 0;
+    void *verify_ctx = NULL;
+    verify_sign_cb_fn verify_sign_fn = NULL;
+
+    ret = (self->quic->verify_certificate_callback_fn)(self->quic->verify_certificate_ctx, cnx,
+                                                       certs, num_certs, &verify_sign_fn, &verify_ctx);
+
+    if (ret == 0) {
+        *verify_sign = verify_sign_callback;
+        *verify_data = malloc(sizeof(picoquic_verify_ctx_t));
+        *((picoquic_verify_ctx_t*)*verify_data) = (picoquic_verify_ctx_t) { verify_ctx, verify_sign_fn };
+    }
+
+    return ret;
+}
+
+int picoquic_enable_custom_verify_certificate_callback(picoquic_quic_t* quic) {
+    picoquic_verify_certificate_t* verifier = NULL;
+    ptls_context_t* ctx = (ptls_context_t*)quic->tls_master_ctx;
+
+    assert(quic->verify_certificate_callback_fn != NULL);
+
+    verifier = (picoquic_verify_certificate_t*)malloc(sizeof(picoquic_verify_certificate_t));
+    if (verifier == NULL) {
+        return PICOQUIC_ERROR_MEMORY;
+    } else {
+        if (ctx->verify_certificate) {
+            free(ctx->verify_certificate);
+        }
+
+        verifier->quic = quic;
+        verifier->cb.cb = verify_certificate_callback;
+        ctx->verify_certificate = &verifier->cb;
+
```

Comment:
Why do I get this warning in the Travis build?
~~~
/home/travis/build/private-octopus/picoquic/picoquic/tls_api.c:496:25: warning: 
      incompatible pointer types assigning to 'int (*)(struct
      st_ptls_verify_certificate_t *, ptls_t *, int (**)(void *, ptls_iovec_t,
      ptls_iovec_t), void **, ptls_iovec_t *, size_t)' (aka 'int (*)(struct
      st_ptls_verify_certificate_t *, struct st_ptls_t *, int (**)(void *,
      struct st_ptls_iovec_t, struct st_ptls_iovec_t), void **, struct
      st_ptls_iovec_t *, unsigned long)') from 'int (ptls_verify_certificate_t
      *, void *, ptls_t *, int (**)(void *, ptls_iovec_t, ptls_iovec_t), void
      **, ptls_iovec_t *, size_t)' (aka 'int (struct
      st_ptls_verify_certificate_t *, void *, struct st_ptls_t *, int (**)(void
      *, struct st_ptls_iovec_t, struct st_ptls_iovec_t), void **, struct
      st_ptls_iovec_t *, unsigned long)') [-Wincompatible-pointer-types]
        verifier->cb.cb = verify_certificate_callback;
~~~